kernel/ls3_fs.c

```diff
@@ -0,0 +1,1058 @@
+    char value[DATA_BLOCK_PAYLOAD];
+    uint64_t next_blockno;
+};
+static_assert(sizeof(struct data_block) == PAGE_CACHE_SIZE);
+
+
+#define MAX_KEYLEN (127)
+struct key_entry { // 144 bytes per struct
+    char key[MAX_KEYLEN+1]; // zero-terminated string, 128 bytes total
+    uint64_t data_len; // 8 bytes
+    uint64_t data_blockno; // 8 bytes
+};
+static_assert(sizeof(struct key_entry) == 144);
+
+#define ENTRIES_PER_KEYBLOCK (56)
+struct key_block {
+    uint64_t older_blockno;
+    struct key_entry entry[ENTRIES_PER_KEYBLOCK];
+    char unused[120]; // padding to ensure key_block is one full block
+};
+static_assert(sizeof(struct key_block) == PAGE_CACHE_SIZE);
+
+#define INVALID_BLOCKNO ((uint64_t)(-1LL))
+
+#define MAX_FREEMAP_BLOCKS (1019)
+#define MAX_FILESYSTEM_BLOCKS (MAX_FREEMAP_BLOCKS * PAGE_CACHE_SIZE * 8LLU)
+struct superblock {
+    uint64_t magic; // for diagnostics / sanity checking
+    uint64_t master; // blocknum of master key_block (first block in chain)
+    uint64_t master_list_len; // length of master key_block chain
+    uint64_t fs_size; // size of storage device, in bytes
+    uint64_t timestamp; // current time
+    uint64_t freemap_block[MAX_FREEMAP_BLOCKS]; // blocknums where freemap is stored
+};
+static_assert(sizeof(struct superblock) == PAGE_CACHE_SIZE);
+
+// need to modify free freemap where things are used initially
+
+struct mount_data {
+    struct superblock *super;
+    struct file *filp;
+
+    // NOTE: freemap is a bitmap used to track status of ONLY the key and data
+    // blocks, The first 32 blocks are always set as USED, in order to reserve
+    // them for superblocks. The next X blocks are always set as USED to reserve
+    // them for storing the freemap itself (at least double the size needed for
+    // a full copy of the freemap).
+    unsigned long *freemap;
+
+    // Freemap is stored in a few blocks (lets say 3 blocks to store around
+    // 100000 bits, for 100000 blocks), but we reserve twice that amount, so X=6
+    // blocks (we could reserve more than this). Initially, the 3 blocks of the
+    // freemap will just go in blocks 32, 33, 34, and the other blocks, 35, 36,
+    // 37 are unused. If part of the freemap becomes dirty, it is put into block
+    // 35. If another part (or the same part) becomes dirty later, it is put
+    // into block 36, etc. In memory, we keep track of the highest blockno used
+    // for the freemap. Once it reaches the end of our reserved region, we erase
+    // the entire region (all 6 blocks), and store the freemap in the first 3
+    // blocks again.
+
+    // Keep track of the highest
+    uint64_t highest_freemap_blockno;
+
+    // freemap_array_status is an in-memory-only bitmap used to track the
+    // status (CLEAN or DIRTY) of the freemap_block array. When one of the
+    // array entries is dirty, it will need to be flushed to the storage device
+    // eventually, along with a new superblock.
+    unsigned long *freemap_array_status;
```

Comment:
won't we need 3 values for this?

Reply:
the idea here is we really just need an int for both the superblock area and the bitmap area to keep track of what is next. When we get to HIGHEST_FREEMAP_BLOCKNO, erase everything